### DISCUSSION TITLE
Multiple Viewports

### DISCUSSION THREAD

**Discussion by VedankNaik** [iTwin/itwinjs-core#3553] [Q&A] [answered]
I've implemented [Multiple Viewports](https://www.itwinjs.org/sandboxes/iTwinPlatform/Multiple%20Viewports) sample. Is it possible to add more than 2 viewports? 
I want 4 viewports in this arrangement
![vp](https://user-images.githubusercontent.com/47364632/165272920-b8a8a64e-3b68-4b65-b49f-7c11b26412cc.png)

**Comment by pmconne** [comment, marked as answer]
The @itwin/core-frontend package supports any number of viewports in any arrangement. e.g., in addition to a grid, you could have viewports floating on top of other viewports, at opposite corners of the page, or whatever else you want.
@NancyMcCallB and/or @bsteinbk can tell you what constraints the ui packages place upon that freedom.

**Reply by bsteinbk** [reply]
@VedankNaik - try updating that sample with the following code.

```
    // Create the content group.
    this._contentGroup = new ContentGroup({
      id: "MultiViewportContentGroup",
      layout: StandardContentLayouts.fourQuadrants,
      contents: [
        {
          id: "MultiViewport1",
          classId: IModelViewportControl,
          applicationData: {
            viewState,
            iModelConnection: connection,
          },
        },
        {
          id: "MultiViewport2",
          classId: IModelViewportControl,
          applicationData: {
            viewState,
            iModelConnection: connection,
          },
        },
        {
          id: "MultiViewport3",
          classId: IModelViewportControl,
          applicationData: {
            viewState,
            iModelConnection: connection,
          },
        },        {
          id: "MultiViewport4",
          classId: IModelViewportControl,
          applicationData: {
            viewState,
            iModelConnection: connection,
          },
        },
      ],
    });
```

You will also have to add the following import to the top of the file.

```
import { StandardContentLayouts } from "@itwin/appui-abstract";
```

**Reply by bsteinbk** [reply]
Also, See [Floating content views](https://www.itwinjs.org/learning/ui/appui-react/contentviews/#floating-content-view) for info on how to set up a more free form content view that can be moved around.

**Reply by VedankNaik** [reply]
Thank you, I was able to get 4 viewports. Also, is it possible to sync all viewports?

**Reply by pmconne** [reply]
Can you be more specific about what aspects of the viewports you want to sync? I assume you don't want 4 viewports displaying identical contents.
[TwoWayViewportFrustumSync](https://www.itwinjs.org/reference/core-frontend/views/twowayviewportfrustumsync/) provides an example of how to sync the viewing volumes of 2 viewports. You ought to be able to use it as the basis for writing something that synchronizes more than 2 viewports.

**Reply by VedankNaik** [reply]
Yes similar to sample which synchronizes 2 viewports I wanted to sync other viewports too.

**Reply by pmconne** [reply]
It should be very easy for you to modify the 2-way sync to sync 4 viewports instead.

**Reply by VedankNaik** [reply]
Following [Multiple Viewports](https://www.itwinjs.org/sandboxes/iTwinPlatform/Multiple%20Viewports) sample I tried the following changes in sync part of code:
`public static connectViewports(selectedVp: Viewport, unselectedVp: Viewport[])
{
    const [vp1, vp2] = [unselectedVp[0], unselectedVp[1]];
    MultiViewportApi.twoWaySync.connect(selectedVp, vp1);
    MultiViewportApi.twoWaySync.connect(selectedVp, vp2);
 }
`

**selectedVp** contains viewport which is currently selected by user, **vp1** and **vp2** contains other viewports. This still syncs two viewports for me. Am I doing it wrong?

**Reply by pmconne** [reply]
You are using [TwoWayViewportSync](https://www.itwinjs.org/reference/core-frontend/views/twowayviewportsync/), which only syncs 2 viewports. If you call `connect` repeatedly, it severs the previous connection.

Look at [the implementation](https://github.com/iTwin/itwinjs-core/blob/master/core/frontend/src/TwoWayViewportSync.ts) of TwoWayViewportSync. It sets up event listeners such that whenever a change occurs in one viewport, the same change gets propagated to the other viewport.
You need to write something very similar. The difference is that whenever a change occurs in one viewport, it must be propagated to the 3 other viewports. Most of the code would remain the same.

Really, TwoWayViewportSync could be generalized and replaced with a MultiViewportSync API accepting a list of 2 or more viewports. If you have success, that would be a nice PR.

**Reply by pmconne** [reply]
Multi-viewport synchronization added in #3614.

**Comment by VedankNaik** [comment]
Thanks for the update @pmconne.

How do I display different models in different viewports? Currently all viewports are displaying same connection. How do I establish connection with different models?

**Reply by pmconne** [reply]
DIfferent *iModels* or different *models*?
I can tell you how to do either using the APIs in @itwin/core-frontend, that's easy.
I suspect you are asking how to do it using the ui packages, which I cannot help with.
Please post a new, more specific question with relevant code etc.

**Reply by VedankNaik** [reply]
Ok I'll post new question. Can you tell me what is difference between imodel and model?

Thanks.

**Reply by pmconne** [reply]
[An iModel is comprised of many models](https://www.itwinjs.org/learning/imodels/#an-imodel-is-comprised-of-many-models).
[A model is comprised of many elements](https://www.itwinjs.org/learning/imodels/#an-imodel-is-comprised-of-many-models).
